### PR TITLE
[build] Bump minimal Dune required version to 2.9

### PR DESCRIPTION
**Pull request by ejgallego**
2.9 has been around for a while and should allow us to:

- generate the opam files
- make progress on https://github.com/coq/coq/pull/14519
- make progress on https://github.com/coq/coq/pull/14241

### CODE REVIEW

**Comment by Alizter** [comment]
To be clear, you are bumping to 2.9 because that is the most recent supported by Debian correct? Otherwise I would have bumped to 3.0 or 3.1.

**Comment by ejgallego** [comment]
> To be clear, you are bumping to 2.9 because that is the most recent supported by Debian correct? Otherwise I would have bumped to 3.0 or 3.1.

Not exactly, I am bumping to 2.9 because I'm more conservative on several fronts, including not willing to depart yet from the 2.x build language.

**Comment by Alizter** [comment]
>  including not willing to depart yet from the 2.x build language.

What is keeping you back?

**Review by Alizter — APPROVED** [review]
I'm fine with this.

@JasonGross I see that Debian has Dune 2.9. Can you double check that this is OK for packaging?

**Comment by Alizter** [comment]
OK merging when CI is green.

File: .github/workflows/ci-macos.yml [hunk 1]
```diff
@@ -21,7 +21,7 @@ jobs:
           opam switch set ocaml-base-compiler.$COMPILER
           eval $(opam env)
           opam update
-          opam install -j "$NJOBS" ocamlfind${FINDLIB_VER} ounit lablgtk3-sourceview3 zarith.1.11 dune.2.8.5
+          opam install -j "$NJOBS" ocamlfind${FINDLIB_VER} ounit lablgtk3-sourceview3 zarith.1.11 dune.2.9.3
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply .github/workflows/ci-macos.yml:24]: Ok we can use that, the fixes in 2.9.2 / 2.9.3 are not imporant for Coq, but this is anyways the OSX package so I didn't think it was important.

**Comment by Alizter** [comment]
@Zimmi48 I backed down from wanting to merge anyway since I didn't form a clear opinion whether or not Debian will be packageable. I'm also not too interested in being assigned here anymore.

File: coq-doc.opam [hunk 2]
```diff
@@ -16,7 +16,7 @@ homepage: "https://coq.inria.fr/"
 doc: "https://coq.github.io/doc/"
 bug-reports: "https://github.com/coq/coq/issues"
 depends: [
-  "dune" {build & >= "2.5.0"}
+  "dune" {>= "2.9"}
```
Comment (SkySkimmer) [inline coq-doc.opam:19]: Was `build` incorrect?

**Review by SkySkimmer — APPROVED** [review]

File: dev/ci/docker/bionic_coq/Dockerfile [hunk 3]
```diff
@@ -46,7 +46,7 @@ ENV COMPILER="4.09.0"
 # Common OPAM packages
 ENV BASE_OPAM="zarith.1.11 ounit2.2.2.3" \
     CI_OPAM="ocamlgraph.1.8.8 yojson.1.7.0 cppo.1.6.8" \
-    BASE_ONLY_OPAM="dune.2.9.1 elpi.1.15.2 stdlib-shims.0.1.0 ocamlfind.1.8.1 odoc.1.5.3"
+    BASE_ONLY_OPAM="dune.2.9.3 elpi.1.15.2 stdlib-shims.0.1.0 ocamlfind.1.8.1 odoc.1.5.3"
```
Comment (SkySkimmer) [inline dev/ci/docker/bionic_coq/Dockerfile:49]: Maybe this should stay on 2.9.1

File: coq-doc.opam (hunk 2 above)
Comment (ejgallego) [reply coq-doc.opam:19]: Until Dune 2.3 it was Ok, later on, the `dune-package` meta-data files became versioned, so indeed, if dune is updated the metadata needs to be regenerated.

This is a pitfall, tho the cache doesn't cache dune itself so it helps.

File: .github/workflows/ci-macos.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ejgallego) [reply .github/workflows/ci-macos.yml:24]: Fixed thanks.

**Comment by ejgallego** [comment]
I think I addressed all of the comments.

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Comment by Zimmi48** [comment]
FTR, the conclusion of the discussion in #16071 (which should have been repeated here) was that actually only Debian unstable has Dune 2.9.1. So this bump might be considered as a departure from our previous policy regarding dependency version constraints. On the other hand, this bump won't cause any trouble for packaging Coq in Debian, we have already bumped lablgtk to a pretty recent version in Coq 8.16, users can build Coq with opam, and if they don't want to do this, building Dune from source is possible.

**Comment by SkySkimmer** [comment]
I see 2.9.1 in testing https://packages.debian.org/testing/dune

**Comment by Zimmi48** [comment]
Yeah, you can do `sed `s/unstable/testing/g` in my comment above. My point was that it was not in Debian stable and that the next Debian stable was not going to be released before the next Coq release.

**Comment by ejgallego** [comment]
I think we need to discuss (and actually document) our policy w.r.t. dependencies. When the "Debian rule" was created the situation was very very different in many fronts.

**Comment by gares** [comment]
w.r.t. dune and Debian I don't think we have a rule agreed upon, just a request by Jason not to do that since this makes building backports (which build and run on a stable system) harder since one has also to backport dune.

I personally think the burden of backporting for debian should not even be visible to Coq devs, but well, since this time it was, we could have just waited.

**Comment by JasonGross** [comment]
> w.r.t. dune and Debian I don't think we have a rule agreed upon, just a request by Jason not to do that since this makes building backports (which build and run on a stable system) harder since one has also to backport dune.

I think there are two separate concerns here: first, I don't think we should ever require a version that is not already packaged on *some* Ubuntu or Debian (even if it's only on testing or unstable).  My selfish motivation here is that I don't want to package new software, it's much harder than backporting.

Separately, there's the question of whether or not Coq master should be buildable from system packages on the latest Ubuntu LTS and Debian stable.  I think this is nice (and minimizes the number of backports I need to do), but it's nowhere near as much of a hassle for me if Coq breaks this policy.

**Comment by ejgallego** [comment]
@JasonGross , some concern here is that it seems to me that OCaml + Coq on Debian is not seeing a lot of manpower compared to other alternatives such as Docker or Nix. And actually that's for good reason (sadly, as I'm a die-hard debian user)

You seem to be spending a lot of cycles on that, would it make sense for you to consider alternatives?

**Pull request merged**